### ISSUE TITLE
[Question] Laplacian loss function

### ISSUE THREAD

**Issue by LucasAlegre** [mklissa/dceo#1]
Hi! Thanks a lot for providing the code, this is really nice work! :)

I would like to ask for some intuition or references for the `neg_loss_fn` shown in the README. You mention that "recent improvements on this representation learning method have lead to the Proper Laplacian", but I have not found a similar loss function in the paper. 

Thanks in advance!

**Comment by mklissa** [comment]
Hi Lucas, thank you for your kind words! 

I probably should be more clear in the README, but the loss we present (and use in our paper) is the one from the [Generalized Laplacian](https://arxiv.org/abs/2107.05545). We make a reference to a fresh new release on the line of work of approximating the Laplacian representation. This new paper is called the [Proper Laplacian Representation Learning](https://arxiv.org/abs/2310.10833) and from preliminary experiments it seems to work even better in terms of approximating the Laplacian eigenfunctions and also seems more stable than what we used in our experiments. However in the Proper Laplacian paper, the focus is not on learning options but rather approximating the Laplacian representation. I hope this helps.

**Comment by LucasAlegre** [comment]
Thanks for the response! I think I can see more clearly now that the code implements the loss in the Generalized Laplacian paper.

The only thing that is a bit confusing to me is how the following part
```
loss += coeff * (dot_product ** 2 - norm_transform(x_norm)  - norm_transform(y_norm)  )
```
approximates Eq. 7 and Eq. 8 in the [Generalized Laplacian paper](https://arxiv.org/abs/2107.05545). Is the `jnp.log1p` trick used for numerical stability?

**Comment by mklissa** [comment]
That is right. Traditionally a normalization is done using the dimensionality of the Laplacian representation (e.g. this [line]( https://github.com/yifan12wu/rl-laplacian/blob/master/rl_lap/agent/laprepr.py#L44) in the original work on the Laplacian eigenfunctions). However, I found this to be insufficient in high dimensional domains and instead resorted to using a different transform on the norms. 

The `jnp.log1p` essentially acts in similar way: norms are encouraged to move away from zero (as seen in Equation 7 of the original Laplacian in RL [paper](https://arxiv.org/pdf/1810.04586.pdf)). But, contrarily to using `norms**2 / lap_dim`, the gradients do not grow with the magnitude of the norms. This really stabilized the representation in my experiments.

**Comment by LucasAlegre** [comment]
Thanks for the clarification! :)

**Comment by leekwoon** [comment]
Hi @mklissa,

Thank you very much for sharing your code. I've been exploring the implementation of Laplacian representation, and I noticed something interesting regarding how negative samples are handled.

In the original implementation, it appears that a single negative sample is generated for computing the negative loss, as seen here: [line](https://github.com/yifan12wu/rl-laplacian/blob/master/rl_lap/agent/laprepr.py#L26).

However, in your version, it seems like two different samples (phi_random_u and phi_random_v) are used for calculating the neg_loss. Could you elaborate on the reasoning behind sampling twice for this calculation? Is there a specific advantage?

Looking forward to your insights. Thank you!
